lxd/acme/acme.go

```diff
@@ -0,0 +1,127 @@
+package acme
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/go-acme/lego/certcrypto"
+	"github.com/go-acme/lego/certificate"
+	"github.com/go-acme/lego/lego"
+	"github.com/go-acme/lego/registration"
+
+	"github.com/lxc/lxd/lxd/state"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+// UpdateACMECertificate updates the ACME certificate.
+func UpdateACMECertificate(s *state.State, provider HTTP01Provider, clustered bool, domain string, email string, caURL string) (*certificate.Resource, error) {
+	// Load the certificate.
+	certInfo, err := util.LoadCert(s.OS.VarDir)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load certificate and key file: %w", err)
+	}
+
+	cert, err := x509.ParseCertificate(certInfo.KeyPair().Certificate[0])
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse certificate: %w", err)
+	}
+
+	issueNewCertificate := true
+
+	// If certificate domain matches the configured domain, renew the cert.
+	if shared.StringInSlice(domain, cert.DNSNames) {
+		issueNewCertificate = false
+	}
+
+	// Check the certificate's expiration date, and renew it if it's only valid for less than 30 days.
+	if !issueNewCertificate && !time.Now().After(cert.NotAfter.Add(-time.Hour*24*30)) {
```

Comment:
You could just add `shared.StringInSlice(domain, cert.DNSNames)` here rather than have the `issueNewCertificate` var - as its somewhat misleading anyway as the `issueNewCertificate` can be `true` and still not issue a certificate due to the expiry date.